book/security.md

```diff
@@ -38,732 +43,781 @@ personalization or show you your private data.
+`POST` requests, but is sent on same-site `POST` or cross-site `GET`
+requests.[^iow-links]
+
+[^iow-links]: Cross-site `GET` requests are also known as "clicking a
+    link", so you can see why this is a smart set of restrictions. The
+    `Strict` version of `SameSite` blocks these too, which can work
+    for some websites but not others.
+    
+To start, let's find a place to store this attribute. I'll modify
+`COOKIE_JAR` to store cookie/parameter pairs:
+
+``` {.python replace=(url/(url%2c%20top_level}
+def request(url, payload=None):
+    if "set-cookie" in headers:
+        params = {}
+        if ";" in headers["set-cookie"]:
+            cookie, rest = headers["set-cookie"].split(";", 1)
+            for param_pair in rest:
+                name, value = param_pair.split("=", 1)
+                params[name.lower()] = value.lower()
+        else:
+            cookie = headers["set-cookie"]
+        COOKIE_JAR[host] = (cookie, params)
 ```
 
-The `hidden` input type basically instructs a browser not to render the
-input area, but to still submit it in the form. In our browser, which
-doesn't support hidden input elements, this will show up as an input
-area with a random string of digits---not ideal, but as long as you
-don't touch it form submission will work fine.
+When sending a cookie in an HTTP request, the browser only sends the
+cookie value, not the parameters:
 
-We'll also need to save the nonce so that we can tell valid from
-invalid nonces. I'm going to store that in a `NONCES` variable, which
-will store one nonce per user. We'll then check both that the nonce is
-valid and also that it is associated with the correct user:[^22]
+``` {.python replace=(url/(url%2c%20top_level}
+def request(url, payload=None):
+    if host in COOKIE_JAR:
+        cookie, params = COOKIE_JAR[host]
+        body += "Cookie: {}\r\n".format(cookie)
+```
 
-[^22]: Otherwise an attacker could get a nonce for *their* account and
-    then use in a CSRF attack for another user's submission.
+Now we can reference the `SameSite` parameter of a cookie. But to
+actually use it, we need to know which site an HTTP request is being
+made from. Let's add a new `top_level` parameter to `request` to track
+that:
 
-``` {.python expected=False}
-def show_comments(username):
+``` {.python}
+def request(url, top_level, payload=None):
     # ...
-    if username:
-        nonce = str(random.random())[2:]
-        NONCES[username] = nonce
 ```
 
-When the form is submitted, we will check the nonce:
+Our browser calls `request` in three places, and we need to send the
+top-level URL in each case. At the top of `load`, it makes the initial
+request to a page. Modify it like so:
 
-``` {.python expected=False}
-def add_entry(params, username):
-    if not check_nonce(params, username):
-        return "Invalid nonce", {}
-    # ...
+``` {.python}
+class Tab:
+    def load(self, url, body=None):
+        headers, body = request(url, self.url, payload=body)
```

Comment:
This behavior restricts navigations from one url to another to skip cookies...I don't think that is right, as I mentioned above.